utils/build-script

```diff
@@ -66,6 +65,480 @@ def call_without_sleeping(command, dry_run=False):
+            args.skip_test_watchos = True
+
+        # --skip-test-ios is merely a shorthand for host and simulator tests.
+        if args.skip_test_ios:
+            args.skip_test_ios_host = True
+            args.skip_test_ios_simulator = True
+        # --skip-test-tvos is merely a shorthand for host and simulator tests.
+        if args.skip_test_tvos:
+            args.skip_test_tvos_host = True
+            args.skip_test_tvos_simulator = True
+        # --skip-test-watchos is merely a shorthand for host and simulator
+        # --tests.
+        if args.skip_test_watchos:
+            args.skip_test_watchos_host = True
+            args.skip_test_watchos_simulator = True
+
+        # --skip-build-{ios,tvos,watchos}-{device,simulator} implies
+        # --skip-test-{ios,tvos,watchos}-{host,simulator}
+        if args.skip_build_ios_device:
+            args.skip_test_ios_host = True
+        if args.skip_build_ios_simulator:
+            args.skip_test_ios_simulator = True
+
+        if args.skip_build_tvos_device:
+            args.skip_test_tvos_host = True
+        if args.skip_build_tvos_simulator:
+            args.skip_test_tvos_simulator = True
+
+        if args.skip_build_watchos_device:
+            args.skip_test_watchos_host = True
+        if args.skip_build_watchos_simulator:
+            args.skip_test_watchos_simulator = True
+
+        if not args.host_test:
+            args.skip_test_ios_host = True
+            args.skip_test_tvos_host = True
+            args.skip_test_watchos_host = True
+
+        if args.build_subdir is None:
+            args.build_subdir = \
+                swift_build_support.workspace.compute_build_subdir(args)
+
+        # Add optional stdlib-deployment-targets
+        if args.android:
+            args.stdlib_deployment_targets.append(
+                StdlibDeploymentTarget.Android.armv7)
+
+    def __init__(self, toolchain, args):
+        self.toolchain = toolchain
+        self.args = args
+
+        self.workspace = swift_build_support.workspace.Workspace(
+            source_root=SWIFT_SOURCE_ROOT,
+            build_root=os.path.join(SWIFT_BUILD_ROOT, args.build_subdir))
+
+    def initialize_runtime_environment(self):
+        """Change the program environment for building."""
+
+        # Set an appropriate default umask.
+        os.umask(0o022)
+
+        # Unset environment variables that might affect how tools behave.
+        for v in [
+                'MAKEFLAGS',
+                'SDKROOT',
+                'MACOSX_DEPLOYMENT_TARGET',
+                'IPHONEOS_DEPLOYMENT_TARGET',
+                'TVOS_DEPLOYMENT_TARGET',
+                'WATCHOS_DEPLOYMENT_TARGET']:
+            os.environ.pop(v, None)
+
+    def build_ninja(self):
+        if not os.path.exists(self.workspace.source_dir("ninja")):
+            diagnostics.fatal(
+                "can't find source directory for ninja "
+                "(tried %s)" % (self.workspace.source_dir("ninja")))
```

Comment:
This check was deferred after (possible) `shell.rmtree(build_dir)`.
We *should* perform every validation before any destructive operation.

Anyway, I'm OK with it for now. `build-script-impl` still have many validations.
When we migrate them, we should rearrange them (including this).